src/write/dwarf.rs

```diff
@@ -129,49 +131,151 @@ pub(crate) mod convert {
+        /// )?;
+        /// # unreachable!()
+        /// # }
+        /// ```
         pub fn from<R: Reader<Offset = usize>>(
-            dwarf: &read::Dwarf<R>,
+            from_dwarf: &read::Dwarf<R>,
             convert_address: &dyn Fn(u64) -> Option<Address>,
         ) -> ConvertResult<Dwarf> {
-            let mut line_strings = LineStringTable::default();
-            let mut strings = StringTable::default();
-            let units = UnitTable::from(dwarf, &mut line_strings, &mut strings, convert_address)?;
+            let mut dwarf = Dwarf::default();
+            let mut convert = dwarf.convert(from_dwarf, None)?;
+            while let Some((mut unit, root_entry)) = convert.read_unit()? {
+                if let Some(convert_program) = unit.read_line_program(None, None)? {
+                    let (program, files) = convert_program.convert_all(convert_address)?;
+                    unit.set_line_program(program, files);
+                }
+                unit.convert_attributes(unit.unit.root(), &root_entry, convert_address)?;
+                while let Some((id, entry)) = unit.read_entry()? {
+                    if id.is_none() {
+                        continue;
+                    }
+                    let id = unit.add_entry(id, &entry);
+                    unit.convert_attributes(id, &entry, convert_address)?;
+                }
+            }
             // TODO: convert the line programs that were not referenced by a unit.
-            let line_programs = Vec::new();
-            Ok(Dwarf {
-                units,
-                line_programs,
-                line_strings,
-                strings,
-            })
+            Ok(dwarf)
+        }
+
+        /// Create a converter for all units in the `.debug_info` section of the given
+        /// DWARF object.
+        ///
+        /// `encoding` applies to the converted units, and may be different from the
+        /// source unit. If `None`, the encoding from the source unit is used.
+        ///
+        /// ## Example
+        ///
+        /// Convert a DWARF section using `convert`.
+        /// See [`ConvertUnit`](crate::write::ConvertUnit) for an example of the unit
+        /// conversion.
+        ///
+        /// ```rust,no_run
+        /// # fn example() -> Result<(), gimli::write::ConvertError> {
+        /// # let loader = |name| -> Result<gimli::EndianSlice<gimli::RunTimeEndian>, gimli::Error> { unimplemented!() };
+        /// let read_dwarf = gimli::read::Dwarf::load(loader)?;
+        /// let mut write_dwarf = gimli::write::Dwarf::new();
+        /// let mut convert = write_dwarf.convert(&read_dwarf, None)?;
+        /// while let Some((mut unit, root_entry)) = convert.read_unit()? {
+        ///     // Now you can convert the root DIE attributes, and other DIEs.
+        /// }
+        /// # unreachable!()
+        /// # }
+        /// ```
+        // TODO: specify `encoding` per unit instead?
```

Comment:
Not sure when you would want to have different units have different encodings in the same set of debug info, but since it is a per-unit property, it makes sense to me to reflect that in these APIs.